### ISSUE TITLE
One request Cancels another request

### ISSUE THREAD

**Issue by intergalactic-overlords** [simoneb/axios-hooks#330]
I've been working on a weird bug for the last few days. Let me explain:

I have a screen for a 'team' of 'operators' that are working today. In this screen, I do 2 separate rest calls with axios hooks. In one, I GET the team data (team name, etc...) and in the other I GET the operators.

I use 'makeUseAxios' like this:
```
const planningAxios = axios.create({ baseURL: API_BASE_URLS.planning })
export const usePlanningApi = makeUseAxios({ axios: planningAxios })
```

My endpoints look like this:
* teams: /api/teams/{teamId}
* operators: /api/teams/{teamId}/operators

Because I have the teamId in te sites url, both these Rest call are executed more or less simultaneously in the component.
For certain teams, I have only a few operators, but in others I have a lot of them. When there are a lot, the 'operators request' takes longer to return a result. 
Here is where things get weird. When it takes longer than a few seconds, the 'operators request' is cancelled. But the weird thing is, it is cancelled by the 'teams request'.

I have stepped through axios-hooks code in dev-tools with breakpoints, and what happens is, at a certain point, the 'teams request' goes into 'withCancelToken' function and does 'cancelOutstandingRequest' and the 'operators request' is cancelled...

Also when I remove the 'teams request' from the page, the 'operators request' works just fine.

I have the same behaviour in both Firefox and Chrome.
But in Chrome devtools, everything (except the cancelled request) returns a 200. But when I look at the rest calls in Firefox, teams has a 304 (not modified) response. (But I've read that Chrome handles caching a different way.)

It seems like the 304 (or whatever happens in Chrome) triggers a 'Cancel' and then everything is cancelled, not only the 304 request, but all other requests as well.

**Workaround:**
I first execute the 'teams request' and then the 'operators request'. Operators request is a manual request now, so I wait for the results of team before executing the operators request (in a useEffect). This way, when 'cancelOutstandingRequest' is triggered, operators request isn't running yet.
It's not ideal, but it's the best solution I can find for now.

I know this is probably an edgecase and not very easy to replicate I guess, but I hope you find a (better) solution for this :-)

**Comment by simoneb** [comment]
Interesting issue. I tried reproducing it but I couldn't, can you try to see if you can reproduce it starting from this repro which as far as I can see works as expected?

https://codesandbox.io/s/axios-hooks-issue-330-r34t0?file=/src/index.js

**Comment by dkershner6** [comment]
I have experienced something different, but similar, and it had to do with the initial loading of the `useAxios` hook with an incomplete url (then changing it to the correct URL).  This would be similar to your `teamId` initially being undefined, then populating (from some other area as a prop).

It doesnt appear that you are populating this separately, but the reason this would cause problems is because it would re-fire the useEffect (essentially a re-render of the hook).  Since you appear to be experiencing a race, as was I, I am guessing something is occurring to change the config (could be anything...headers?).

My workaround was to always set the initial config to an empty string (`''`) (null throws error),  `manual: true`, and run the refetch (the first fetch in this case) with the proper URL config as a param (override).

**Comment by simoneb** [comment]
@dkershner6 if I understand your issue correctly, you don't want to fire the request until you have the url available, which is something that you don't have initially.

If that is the case, and assuming for instance that you get the url from a property of the component where you're using the hook, you can simply do something like:

```jsx
function MyComponent({ url }) {
  useAxios(url, { manual: !url })
}
```

In other words, you prevent the request from executing until you have a url.

**Comment by dkershner6** [comment]
@simoneb , That's an interesting thought: change the manual setting instead of calling the callback. This would also re-Render, but all logic would be handled in the useEffect and would work.

I like the explicitness of the callback, since manual isn't an accurate name at that point, but I won't rule it out in simple use cases. I also think it's worth a thought to try and get the callback to work in this case, since it seems like it would be a common pattern. I couldn't figure it out completely, but my PR will help.

**Comment by simoneb** [comment]
@dkershner6 can we close this issue?

**Comment by dkershner6** [comment]
I think that the likelihood of this being solved by #358 are pretty good, or at least helped by it.  If we combine this with the ability to:

- Use blank string as a blank slate.
- Use `manual` as an on/off switch.

I think we can then likely close this ticket.  Might be worth adding something to examples or docs, though, as I definitely interpreted things differently.
